src/natural_number_set.cpp

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+#include <boost/dynamic_bitset.hpp>
+
+#include "ann_exception.h"
+#include "natural_number_set.h"
+
+namespace diskann {
+  template<typename T>
+  natural_number_set<T>::natural_number_set()
+      : _values_bitset(std::make_unique<boost::dynamic_bitset<>>())
+  {
+  }
+
+  template<typename T>
+  bool natural_number_set<T>::is_empty() const
+  {
+      return _values_vector.empty();
+  }
+
+  template<typename T>
+  void natural_number_set<T>::reserve(size_t count)
+  {
+      _values_vector.reserve(count);
+      _values_bitset->reserve(count);
+  }
+
+  template<typename T>
+  void natural_number_set<T>::insert(T id)
```

Comment:
Is this data structure thread safe? Can multiple threads insert and pop safely at the same time?

Reply:
It's not, same as the `robin_set` it replaced for `_empty_slots`. The general rule for `std` containers is that multiple read-only operations are OK, but there can be only one writer at one point.

I'll add an explicit comment about thread-safety.